sycl/source/backend/Intel_level0.cpp

```diff
@@ -0,0 +1,96 @@
+//==------- Intel_level0.cpp - SYCL Level-Zero backend ---------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl.hpp>
+#include <detail/platform_impl.hpp>
+#include <detail/plugin.hpp>
+#include <detail/program_impl.hpp>
+#include <detail/queue_impl.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace level0 {
+using namespace detail;
+
+// Get the L0 plugin.
+static const plugin &getPlugin() {
+  static const plugin *L0Plugin = nullptr;
+  if (L0Plugin)
+    return *L0Plugin;
```

Comment:
Is it possible that this code is run in multiple threads?
If so, some blocking will be required or an atomic pointer should be employed.

Reply:
It is possible. But this code is thread safe because L0Plugin is a static variable.  Quote from C++11 standard, I have highlighted the part about multi-threaded environment:

> The zero-initialization (8.5) of all block-scope variables with static storage duration (3.7.1) or thread storage
> duration (3.7.2) is performed before any other initialization takes place. Constant initialization (3.6.2) of a
> block-scope entity with static storage duration, if applicable, is performed before its block is first entered.
> An implementation is permitted to perform early initialization of other block-scope variables with static or
> thread storage duration under the same conditions that an implementation is permitted to statically initialize
> a variable with static or thread storage duration in namespace scope (3.6.2). Otherwise such a variable is
> initialized the first time control passes through its declaration; such a variable is considered initialized upon
> the completion of its initialization. If the initialization exits by throwing an exception, the initialization
> is not complete, so it will be tried again the next time control enters the declaration. **If control enters
> the declaration concurrently while the variable is being initialized, the concurrent execution shall wait for
> completion of the initialization. (89)** If control re-enters the declaration recursively while the variable is being
> initialized, the behavior is undefined. 
> 
>  **89) The implementation must not introduce any deadlock around execution of the initializer**